### PR TITLE
refactor firmware update

### LINKED ISSUE #702: Suite does not recognize device in bootloder mode (older firmware versions)
Trezor One

Steps to reproduce: 
1. Connect a device with older fw (1.5.2 / 1.6.0) in bootloader mode
Result: Suite keeps saying Connect Trezor to Continue

If I connect the device normally I get "Switch to bootloader screen", but it is impossible to switch to bootloader.

Starting from the fw version 1.6.3 it is okay.

### LINKED ISSUE #862: Wallet creation stuck on "Get the latest firmware" with new device
Suite version: Pipeline 99357995, commit ba7c684a
FW after Suite install: 2.1.1
Operating System: NixOS 19.09.1279.5f506b95f9f
KDE Plasma Version: 5.16.5
KDE Frameworks Version: 5.61.0
Qt Version: 5.12.3
Kernel Version: 4.19.84
OS Type: 64-bit
Processors: 8 × Intel® Core™ i5-8250U CPU @ 1.60GHz
Memory: 15,5 GiB of RAM

When I try to create a new wallet with a fresh device (the device had hologram before the creating of the wallet), the onboarding installs firmware and then it's stuck on "Get the latest firmware". When I click on Back, I'm presented with a screen with "According to your decision in a previous step, this was supposed to be a fresh device. But we were able to detect already installed firmware on it., which cycles the whole process. **THIS DOESNT HAPPEN WHEN I WIPE THE DEVICE** with trezorcrl wipe-device -b, then I get the correct screen with Continue (last screen). Its also seems that when I use the fresh device, Suite installs only 2.1.1, but then when I wipe the device, the firmware installed is 2.1.8. There are some fresh Trezors T on QA, if you need more, Erik can provide some.
I'm using $ rm -rf /home/**user**/.config/@trezor/suite-desktop/IndexedDB to clear indexedDB

Maybe check the bootloader version of the devices on QA, please.

Picks bellow 

Screenshot_20191129_163525

Screenshot_20191129_164212

After trezorctl wipe-device -b
Screenshot_20191129_164925

### PR DESCRIPTION
**Pull request by mroz22**
- refactored firmware update logic
- implemented basic (no design) update flow to firmware application modal

- [ ] resolves #702 (probably, testing would be nice) 
- [ ] resolves #862 (probably, testing would be nice)

### CODE REVIEW

File: packages/suite/src/views/settings/device/Container.ts [hunk 1]
```diff
@@ -10,6 +10,7 @@ import {
 } from '@suite-actions/deviceSettingsActions';
 // todo: device management actions
 import * as backupActions from '@suite-actions/backupActions';
+// import * as firmwareActions from '@suite-actions/firmwareActions';
```
Comment (szymonlesisz) [inline packages/suite/src/views/settings/device/Container.ts]: remove

File: packages/suite/src/views/onboarding/steps/Firmware/Container.ts [hunk 2]
```diff
@@ -11,6 +11,7 @@ const mapStateToProps = (state: AppState) => ({
     deviceCall: state.onboarding.deviceCall,
     device: state.suite.device,
     firmwareUpdate: state.firmware,
+    // locks: state.suite.locks,
```
Comment (szymonlesisz) [inline packages/suite/src/views/onboarding/steps/Firmware/Container.ts]: remove

File: packages/suite/src/views/settings/device/index.tsx [hunk 3]
```diff
@@ -148,7 +149,8 @@ const Settings = ({
                             title={<Translation>{messages.TR_FIRMWARE_VERSION}</Translation>}
                             description={
                                 <Translation
-                                    values={{ version: getFwVersion(device) }}
+                                    values={{ version: '1.2.3' }}
```
Comment (szymonlesisz) [inline packages/suite/src/views/settings/device/index.tsx]: remove hardcoded

File: packages/suite/src/views/suite/firmware/index.tsx [hunk 4]
```diff
@@ -1,384 +1,393 @@
...
-export default connect(mapStateToProps, mapDispatchToProps)(FirmwareUpdate);
+import React from 'react';
+
+export default () => (
+    <>
+        TODO: deprecated page, will be handled in settings, but I need to solve 'application' modal
+        logic first
+    </>
+);
+
```
Comment (szymonlesisz) [inline packages/suite/src/views/suite/firmware/index.tsx]: remove this file

**Comment by mroz22** [comment]
I am going to continue this and finish entire fw update modal here after #976 is merged

File: packages/suite/src/actions/firmware/firmwareActions.ts [hunk 5]
```diff
@@ -2,32 +2,55 @@ import TrezorConnect from 'trezor-connect';
...
+            payload: 'device must be connected in bootloader mode',
+        });
+        return;
     }
 
-    dispatch({ type: FIRMWARE.SET_UPDATE_STATUS, payload: 'started' });
-    dispatch({ type: FIRMWARE.SET_UPDATE_STATUS, payload: 'downloading' });
+    const model = device.features.major_version;
+
+    dispatch(lockUI(true));
```
Comment (szymonlesisz) [inline packages/suite/src/actions/firmware/firmwareActions.ts]: use local state in component

File: packages/suite/src/actions/firmware/firmwareActions.ts [hunk 6]
```diff
@@ -62,18 +87,5 @@ export const firmwareUpdate = () => async (dispatch: Dispatch, getState: GetStat
         return dispatch(lockUI(false));
```
Comment (szymonlesisz) [inline packages/suite/src/actions/firmware/firmwareActions.ts]: use local state in component

File: packages/suite/src/actions/suite/__tests__/suiteActions.test.ts [hunk 7]
```diff
@@ -64,6 +64,15 @@ jest.mock('trezor-connect', () => {
...
 
+jest.mock('next/router', () => {
+    return {
+        __esModule: true, // this property makes it work
+        default: {
+            back: () => {},
+        },
+    };
+});
+
```
Comment (szymonlesisz) [inline packages/suite/src/actions/suite/__tests__/suiteActions.test.ts:75]: ?

File: packages/suite/src/middlewares/firmware/firmwareMiddleware.ts [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+        case SUITE.SELECT_DEVICE:
+            if (
+                action.payload &&
+                action.payload.features &&
+                ['unplug', 'wait-for-reboot'].includes(status) &&
+                action.payload
+            ) {
+                if (action.payload.firmware === 'valid') {
+                    api.dispatch(firmwareActions.setStatus('done'));
+                    api.dispatch(suiteActions.lockUI(false));
```
Comment (szymonlesisz) [inline packages/suite/src/middlewares/firmware/firmwareMiddleware.ts]: use local state in component

File: packages/suite/src/middlewares/firmware/firmwareMiddleware.ts [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+                } else if (['outdated', 'required'].includes(action.payload.firmware)) {
+                    api.dispatch(firmwareActions.setStatus('partially-done'));
+                    api.dispatch(suiteActions.lockUI(false));
```
Comment (szymonlesisz) [inline packages/suite/src/middlewares/firmware/firmwareMiddleware.ts]: use local state in component

File: packages/suite/src/reducers/firmware/firmwareReducer.ts [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+    return produce(state, draft => {
+        switch (action.type) {
+            case FIRMWARE.SET_UPDATE_STATUS:
+                draft.status = action.payload;
+                break;
+            case FIRMWARE.SET_ERROR:
+                draft.status = 'error';
+                draft.error = action.payload;
+                break;
+            case UI.REQUEST_BUTTON:
```
Comment (szymonlesisz) [inline packages/suite/src/reducers/firmware/firmwareReducer.ts]: use solution from set pin t1

File: packages/suite/src/views/settings/device/index.tsx [hunk 11]
```diff
@@ -57,12 +58,13 @@ const Settings = ({
         setLabel(device.label);
     }, [device]);
 
-    if (!device || !device.features) {
-        return null;
+    // any idea how to write this in a sane way
+    let features: any;
```
Comment (szymonlesisz) [inline packages/suite/src/views/settings/device/index.tsx]: ?

**Review by szymonlesisz — APPROVED** [review]

**Pull request merged**